supertokens/supertokens-node
# test: Adds userrole tests

## Summary of change

Adds recipe function tests

## Related issues
https://github.com/supertokens/supertokens-node/pull/324

## Test Plan
- addRoleToUserTests
  - add a role to a user and check that the user has the role
  - add a the same role to a user twice
  -  add an unknown role to the user
- Config test
- createNewRoleOrAddPermissions test
   - create a new role and check the output
   - create the same role twice
   - create a role with permissions
   - create a role with permissions, add new permissions to the role
   - create a role with duplicate permissions  
- deleteRole
  - create roles, add them to a user, delete one of the roles
  -  delete a role that does not exist
- getPermissionsForRole:
  - create a role with permissions, and check that the role has those permissions
  - get permissions for an unknown role
- getRolesForUser
  - create roles, add them to a user check that the user has the roles
- getRolesThatHavePermission
   - create multiple roles with the same permission. call the function and check for the output
   - get roles for unknown permission
- getUsersForRole
  - create a role, add it, multiple users, and retrieve the users who have the role
  - get users for an unknown role
- removePermissionsFromRole
  - create a role with permissions,  remove one permission from the role, check that it is removed
  - remove permissions from an unknown role
- removeUserRole 
  - create a role, add it to a user, remove it from the user and check that it is removed
  - remove a role the user does not have
  - remove an unknown role from a user 

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   ~~[ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)~~
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   ~~[ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that~~
-   [x] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [x] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-
